torch/distributed/_sharded_tensor/api.py

```diff
@@ -258,6 +258,14 @@ def gather(
             out (:class `torch.Tensor`, optional): The output full tensor.
                 Must to be provided ONLY on ``dst`` rank.
                 Default: ``None``
+
+    .. note:: For NCCL-based processed groups, internal tensor representations
+        of objects must be moved to the GPU device before communication takes
+        place. In this case, the device used is given by
+        ``torch.cuda.current_device()`` and it is the user's responsiblity to
+        ensure that this is set so that each rank has an individual GPU, via
+        ``torch.cuda.set_device()``
+
```

Comment:
I think we should put these docs in the ShardedTensor constructor docs. Essentially if you use ShardedTensor you should always do this irrespective of which API you use.

Users don't use the ShardedTensor constructor typically but we could have a sharded_tensor.rst mentioning this in our public docs in the future.